### PR TITLE
Add Emscripten Build Instructions

### PR DESCRIPTION
**Pull request by mcbarton**
# Description

Please include a summary of changes, motivation and context for this PR.

This PR will add instructions on how to build a wasm version of CppInterOp and xeus-cpp.

Fixes # (issue)

## Type of change

Please tick all options which are relevant.

- [ ] Bug fix
- [ ] New feature
- [x] Requires documentation updates

## Testing

Please describe the test(s) that you added and ran to verify your changes.

## Checklist

- [x] I have read the contribution guide recently

### CODE REVIEW

File: Emscripten-build-instructions.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+python -m pip install jupyterlite-xeus
+jupyter lite build --XeusAddon.prefix=$PREFIX
+```
+
+Once the Jupyter Lite site has built you can test the website locally  
+by executing
+
+```bash
+jupyter lite serve --XeusAddon.prefix=$PREFIX
+```
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:167]: I don't think there is any need to paste the xeus-cpp build again. Just provide a link from xeus-cpp's master for the same. This would ensure we have one place to go to when interested and the changes also need to be done at one place only.

File: Emscripten-build-instructions.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+```bash
+mkdir CppInterOp-wasm
+```
+
+Now move into this directory using the following command
+
+```bash
+cd ./CppInterOp-wasm
+```
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:20]: Name for a build directory is weird. Just use build.

Comment (anutosh491) [reply Emscripten-build-instructions.md:20]: Also order looks weird 

1) setup emsdk first
2) any cloning that's required
3) create the build directory
4) The build configs using cmake
5) make step

File: Emscripten-build-instructions.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+emmake make clang-repl -j $(nproc --all)
+emmake make lld -j $(nproc --all)
+```
+
+Once this finishes building we need to take note of where we built our  
+llvm build. This can be done by executing the following
+
+```bash
+export LLVM_BUILD_DIR=$PWD
+```
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:86]: This is okay but in the same place ....
**I would really like to make it as easy as possible for the user to try out**

Building LLVM against emscripten would itself take greater an 1 hour. 
Now that we're using the same configs as emscripten-forge, I would just do 
1) micromamba create an environment
2) micromamba install llvm -c https://repo.mamba.pm/emscripten-forge
3) use env prefix as `llvm_build_dir`, so basically LLVM_DIR = prefix/lib/cmake/llvm 

This would get the above task done in 2 mins. 

But as you've done ... for the curious user, we also provide a way to build llvm so that they can play around with configs if interested.

File: Emscripten-build-instructions.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+```bash
+cd ../../CppInterOp/
+mkdir build
+cd ./build/
+emcmake cmake -DCMAKE_BUILD_TYPE=Release    \
+                -DUSE_CLING=OFF                             \
+                -DUSE_REPL=ON                               \
+                -DLLVM_DIR=$LLVM_BUILD_DIR/lib/cmake/llvm      \
+                -DLLD_DIR=$LLVM_BUILD_DIR     \
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:99]: lld should have its own dir `LLVM_BUILD_DIR/lib/cmake/lld`

File: Emscripten-build-instructions.md (hunk 1 above)
Comment (anutosh491) [reply Emscripten-build-instructions.md:167]: We should just mention if cppinterop is built from source then we need to provide this config 
`-DCppInterOp_DIR="$CPPINTEROP_BUILD_DIR/lib/cmake/CppInterOp"  \`

File: Emscripten-build-instructions.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                ../
+emmake make -j $(nproc --all)
+```bash
+
+Once this finishes building we need to take note of where we built our  
+llvm build. This can be done by executing the following
+
+```bash
+export CPPINTEROP_BUILD_DIR=$PWD
+```
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:113]: Again .... we should provide a way to build cppinterop from source, but these would be the exact same configs on emscripten-forge, hence it would just be easy for anyone to fetch it from there, rather than building the whole toolchain.

File: Emscripten-build-instructions.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+```bash
+./emsdk/emsdk activate 3.1.45
+source ./emsdk/emsdk_env.sh
+```bash
+
+Now clone the 19.x release of the LLVM project repository 
+and CppInterOp
+
+```bash
+git clone --depth=1 --branch release/19.x https://github.com/llvm/llvm-project.git
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:47]: image

This looks weird in preview.

**Comment by anutosh491** [comment]
Overall formatting looks a bit weird, I don't know why that happens but everything seems left aligned.
Can't we have the docs to fit the page ?


image

File: Emscripten-build-instructions.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                        -DLLVM_ENABLE_LIBXML2=OFF \
+                        -DCLANG_ENABLE_STATIC_ANALYZER=OFF \
+                        -DCLANG_ENABLE_ARCMT=OFF \
+                        -DCLANG_ENABLE_BOOTSTRAP=OFF \
+                        -DCMAKE_CXX_FLAGS="-Dwait4=__syscall_wait4" \
+                        ../llvm
+emmake make clang -j $(nproc --all)
+emmake make clang-repl -j $(nproc --all)
+emmake make lld -j $(nproc --all)
+```
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:79]: We have `cmake` in our `envoronment-wasm-build.yml` and I think we that should be setup and activate otherwise the emcmake cmake would fail !

Comment (mcbarton) [reply Emscripten-build-instructions.md:79]: It will not fail. This is evident through the ci since it just builds our llvm with the non conda provided cmake.We are going to assume that for these build instructions

**Comment by mcbarton** [comment]
@anutosh491 I am aware if the weird formatting and will fix before it gets merged when we are happy with the contents

File: Emscripten-build-instructions.md (hunk 5 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:113]: These instructions are here not just for those who want to fownload but also for those who want to help wuth development. I will add a link to the emscripten forge packages if you have one though

File: Emscripten-build-instructions.md (hunk 4 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:99]: Will fix

File: Emscripten-build-instructions.md (hunk 3 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:86]: We are not quite the same (not just yet anyway 😉). These are there to provide the imquistitive user too in case they want to play with some different flags. Gives them a starting point. These instructions are for developers as well as users

File: Emscripten-build-instructions.md (hunk 2 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:20]: This is a top levell build directory to keep everything contained if the development build goes wrong. For those who simple want to experiment with it then I shalll point then tk the emscripten forge package

File: Emscripten-build-instructions.md (hunk 1 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:167]: I plan to keep these instructions here rather than point to xeus-cpp instructions. This is because the instructions are slightly different if testing xeus-cpp against a CppInterOp built from source.

File: Emscripten-build-instructions.md (hunk 7 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:79]: We don't really need envoronment-wasm-build.yml since we can use the systems conda. See llvm build in the ci which does just that.

Comment (anutosh491) [reply Emscripten-build-instructions.md:79]: Hmmm wait but we should atleast ask the user to get cmake from conda-forge using micromamba correct. 
I think you're right and the CI won't need it but if you do a micromamba install, I don't think you get cmake in your base env by default. 

Atleast I get a failure locally when I run a cmake command without having cmake installed

Comment (mcbarton) [reply Emscripten-build-instructions.md:79]: I'll add it for now as I'm not too fussed about this.

Comment (mcbarton) [reply Emscripten-build-instructions.md:79]: Just a note though that of course you'll get an error if you don't have cmake installed locally before executing the command.

**Review by alexander-penev — APPROVED** [review]

**Comment by anutosh491** [comment]
I shall review this soon !

File: Emscripten-build-instructions.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+Once this finishes building we need to take note of where we built our llvm build. This can be done by executing the following
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:80]: image

Formatting error !

File: Emscripten-build-instructions.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+                        -DLLVM_ENABLE_ZSTD=OFF \
+                        -DLLVM_ENABLE_LIBXML2=OFF \
+                        -DCLANG_ENABLE_STATIC_ANALYZER=OFF \
+                        -DCLANG_ENABLE_ARCMT=OFF \
+                        -DCLANG_ENABLE_BOOTSTRAP=OFF \
+                        -DCMAKE_CXX_FLAGS="-Dwait4=__syscall_wait4" \
+                        ../llvm
+emmake make clang -j $(nproc --all)
+emmake make clang-repl -j $(nproc --all)
+emmake make lld -j $(nproc --all)
```
Comment (anutosh491) [inline Emscripten-build-instructions.md:77]: I am not fully convinced here (not sure it matters in the build) but 
1) if we use `LLVM_ENABLE_PROJECTS="clang;lld"` that means we want to build all static libraries in this cases coming out of clang and lld
2) Our focus is not to get clang-repl (the approach we use to get clang-repl working in the browser doesn't need `clang-repl.js` or `clang-repl.wasm` rather everything works through `libclangInterpreter.a`)
3) Even if we want to build clang-repl (which has dependency on lldwasm and lldcommon, I guess this would invoke "make lld" already, as required to get clang-repl ..... which in turn makes the below command redundant.

Basically what I am saying is we need a simple `emmake make -j $(nproc --all)` here and nothing else. I had raised a similar point for the CI not sure if that was addressed.

File: Emscripten-build-instructions.md (hunk 8 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:80]: @anutosh491 Can you eloborate? I'm not  sure what the formstting error is.

File: Emscripten-build-instructions.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+micromamba activate xeus-lite-host
+python -m pip install jupyterlite-xeus
+jupyter lite build --XeusAddon.prefix=$PREFIX
+```
+
+Once the Jupyter Lite site has built you can test the website locally  by executing
+
+```bash
+jupyter lite serve --XeusAddon.prefix=$PREFIX
+```
```
[the remark this answers is not in the document]
Comment (mcbarton) [reply Emscripten-build-instructions.md:159]: @anutosh491 I'll wait until you make the change to get xeus-cpp-lite working. I'll change accordingly if necessary.

File: Emscripten-build-instructions.md (hunk 9 above)
Comment (vgvassilev) [reply Emscripten-build-instructions.md:77]: If we just need `libclangInterpreter.a` we should call that -- it should be a target : `make clangInterpreter` or something like that...

File: Emscripten-build-instructions.md (hunk 8 above)
Comment (anutosh491) [reply Emscripten-build-instructions.md:80]: Open the preview and you would realized what I am talking about !

File: Emscripten-build-instructions.md (hunk 9 above)
Comment (anutosh491) [reply Emscripten-build-instructions.md:77]: Ahhh sorry I meant, we need libclangIntepreter.a (and all its respective transitive dependencies like lldwasm, lldcommon from lld ,  clangast, clangbasic etc from clang and llvmsupport etc from llvm)

I think just doing `emmake make clang -j $(nproc --all)` would get the job done

File: Emscripten-build-instructions.md (hunk 8 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:79]: @anutosh491 I think this is fixed now.

File: Emscripten-build-instructions.md (hunk 9 above)
Comment (mcbarton) [reply Emscripten-build-instructions.md:76]: I think this is fixed now. I removed the clang-repl target, leaving just the clang and lld ones.

**Review by anutosh491 — APPROVED** [review]
xeus-cpp-lite build instructions need to be edited but otherwise looks good. Merging after CI passes.
